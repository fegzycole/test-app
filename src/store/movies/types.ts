export const SET_MOVIES = 'SET_MOVIES';
export const SET_ERROR = 'SET_ERROR';

export interface MovieData {
    imdbID: string;
    Title: string;
    Year: string;
    Type: string;
    Poster: string;
}

export interface MoviesState {
    movies?: MovieData[];
    fetchingMovies: boolean;
    error: string;
}

export interface SetMovies {
    type: typeof SET_MOVIES;
    payload: MovieData[];
}

export interface SetError {
    type: typeof SET_ERROR;
    payload: string;
}

export type MoviesActionTypes = SetMovies | SetError;
