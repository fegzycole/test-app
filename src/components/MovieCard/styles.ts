import Typography from '@mui/material/Typography';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import Grid from '@mui/material/Grid';
import { styled } from '@mui/system';

export const StyledGrid = styled(Grid, {
    name: 'StytledGrid',
    slot: 'Wrapper',
})``;

export const StyledCard = styled(Card, {
    name: 'StyledCard',
    slot: 'Wrapper',
})`
    width: 100%;
    position: relative;
`;

export const StyledCardContent = styled(CardContent, {
    name: 'StyledCardContent',
})`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    height: 60px;
    text-transform: uppercase;
`;

export const StyledTypography = styled(Typography, {
    name: 'StyledButton',
})`
    font-size: 13px;
    font-weight: 600;
    color: #000;
    line-height: 18px;
    text-align: left;
`;
