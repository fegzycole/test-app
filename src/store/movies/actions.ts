import { Dispatch } from 'react';
import { MovieData, SetMovies, SET_MOVIES, SetError, SET_ERROR } from './types';
import axios from 'axios';

const apiKey = process.env.REACT_APP_API_KEY;
const apiUrl = process.env.REACT_APP_API_URL;

export const setMovies = (payload: MovieData[]): SetMovies => {
    return {
        type: SET_MOVIES,
        payload,
    };
};

export const setError = (payload: string): SetError => {
    return {
        type: SET_ERROR,
        payload,
    };
};

export const getMovies = (
    searchText: string,
): ((dispatch: Dispatch<{ type: string; payload?: unknown }>) => Promise<void>) => {
    return async (dispatch) => {
        dispatch(setError(''));

        try {
            const request = await axios.get(`${apiUrl}?s=${searchText}&apikey=${apiKey}`);

            const responseData = request.data.Search;

            const movies: MovieData[] = responseData ? (responseData as MovieData[]) : [];

            dispatch(setMovies(movies));
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
            dispatch(setError('An error occurred, check your internet and try again'));
        }
    };
};
