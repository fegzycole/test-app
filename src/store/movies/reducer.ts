import { MoviesActionTypes, MoviesState, SET_MOVIES, SET_ERROR } from './types';

const initialState: MoviesState = {
    movies: undefined,
    fetchingMovies: false,
    error: '',
};

const moviesReducer = (state: MoviesState = initialState, action: MoviesActionTypes) => {
    switch (action.type) {
        case SET_MOVIES:
            return {
                ...state,
                movies: action.payload,
            };
        case SET_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        default:
            return state;
    }
};

export default moviesReducer;
