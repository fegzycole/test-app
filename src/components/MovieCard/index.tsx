import React from 'react';
import CardMedia from '@mui/material/CardMedia';
import { MovieData } from '../../store/movies/types';
import { StyledGrid, StyledCard, StyledCardContent, StyledTypography } from './styles';

export interface MovieCardProps {
    movie: MovieData;
}

export const MovieCard: React.FC<MovieCardProps> = ({ movie }) => {
    const { Title, Poster } = movie;

    return (
        <StyledGrid container item xs={10} md={4} lg={3}>
            <StyledCard>
                <CardMedia component="img" height="250" image={Poster} alt="movie-poster" />
                <StyledCardContent>
                    <StyledTypography>{Title}</StyledTypography>
                </StyledCardContent>
            </StyledCard>
        </StyledGrid>
    );
};

export default MovieCard;
