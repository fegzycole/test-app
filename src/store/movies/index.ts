import * as Types from './types';
import moviesReducer from './reducer';
import * as Actions from './actions';

export { Types, moviesReducer, Actions };
