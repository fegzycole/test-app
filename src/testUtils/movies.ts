import { MovieData } from '../store/movies/types';

export const mockMovie: MovieData = {
    Title: 'Mock Title',
    Type: 'Mock Type',
    imdbID: 'Mock ID',
    Year: '2022',
    Poster: 'Mock Poster',
};

export const mockMovies: MovieData[] = [mockMovie];
