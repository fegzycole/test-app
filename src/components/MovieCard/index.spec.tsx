import { cleanup, render } from '@testing-library/react';
import { mockMovie } from '../../testUtils/movies';
import MovieCard, { MovieCardProps } from './index';

const props: MovieCardProps = {
    movie: mockMovie,
};

describe('<MovieCard />', () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });

    it('should render correctly', () => {
        const view = render(<MovieCard {...props} />);

        expect(view.container).toMatchSnapshot();
    });
});
