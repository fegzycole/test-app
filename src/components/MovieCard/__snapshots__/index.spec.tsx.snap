// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> should render correctly 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-1557hwu-MuiGrid-root-StytledGrid-wrapper"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-qegifv-MuiPaper-root-MuiCard-root-StyledCard-wrapper"
    >
      <img
        alt="movie-poster"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        height="250"
        src="Mock Poster"
      />
      <div
        class="MuiCardContent-root css-1189jzg-MuiCardContent-root-StyledCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1qhct6m-MuiTypography-root-StyledButton-root"
        >
          Mock Title
        </p>
      </div>
    </div>
  </div>
</div>
`;
